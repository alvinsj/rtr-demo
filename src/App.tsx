import { useEffect } from 'react'

import AuthContext from '@/contexts/AuthContext'
import { getPKCEStatus } from '@/utils/auth'
import { deleteStateCookie } from '@/utils/stateCookie'
import { AuthStage } from '@/types'

import LoginButton from '@/components/LoginButton'

import useAuthContextValue from '@/hooks/useAuthContextValue'
import useGetAccessToken from '@/hooks/useGetAccessToken'

import s from './App.module.css'
import LogoutButton from './components/LogoutButton'
import { getAuthStage } from './utils/authStage'

function clearSearchParams() {
  const url = new URL(window.location.href) // Create a URL object from the current window location
  url.search = '' // Clear the search parameters
  window.history.replaceState({}, document.title, url.toString()) // Update the URL without reloading the page
}

function App() {
  const params = new URLSearchParams(window.location.search)
  const state = params.get('state')
  const code = params.get('code')
  const { codeVerifier } = getPKCEStatus(state)
  const {
    isLoading, error, tokens,
    getATWithAuthCode, getATWithRefreshToken
  } = useGetAccessToken()

  const authContext = useAuthContextValue(tokens)

  const authStage = getAuthStage({
    state, code, codeVerifier,
    accessToken: authContext.accessToken,
    refreshToken: authContext.refreshToken,
  })

  useEffect(() => {
    if (authStage.stage === AuthStage.AFTER_AUTH_CODE) {
      deleteStateCookie(authStage.state)
      clearSearchParams()
    } else if (authStage.stage === AuthStage.BEFORE_REFRESH_TOKEN) {
      getATWithRefreshToken(authStage.refreshToken)
    } else if (authStage.stage === AuthStage.BEFORE_AUTH_CODE) {
      getATWithAuthCode(authStage.state, authStage.code, authStage.codeVerifier)
    }
    // once on mount only
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [authStage.stage])

  const statuses = {
    accessToken: authContext.accessToken,
    refreshToken: authContext.refreshToken,
    state,
    code,
    codeVerifier,
  }

  return (
    <AuthContext.Provider value={authContext}>
      <main className={s['app']}>
        {authStage.stage === AuthStage.LOGGED_IN
          ? <LogoutButton className={s['app-loginBtn']} />
          : <LoginButton className={s['app-loginBtn']} />
        }
        <h1>Stage: {authStage.stage}</h1>
        <pre>
          {!isLoading && <>
            <table>
              <tbody>
                {Object.entries(statuses).filter(([, v]) => !!v).map(([key, value]) => (
                  <tr key={key}>
                    <td className='debug-itemName'>{key}</td>
                    <td className="debug-longText">{`${value}`}</td>
                  </tr>
                ))}
              </tbody>
            </table>
          </>
          }
          {status && status.map(s => <div key={s}>{s}</div>)}
          {error && <div className='debug-error'>{error}</div>}
        </pre>
      </main>
    </AuthContext.Provider>
  )
}

export default App
