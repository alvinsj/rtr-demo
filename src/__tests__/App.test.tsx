
import { test, expect, describe } from 'vitest'
import { render } from '@testing-library/react'
import App from '../App'

test('runs vitest', () => {
  expect(1).toBe(1)
})

describe('App', () => {
  test('renders with snapshot', () => {
    const screen = render(<App />)
    expect(screen.container.children[0]).toMatchSnapshot()
  })
})




