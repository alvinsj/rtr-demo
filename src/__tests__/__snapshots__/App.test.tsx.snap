// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders with snapshot 1`] = `
<main
  class="app"
>
  <div
    class="app-loginBtn"
  >
    <span
      class="error"
    />
    <button
      type="submit"
    >
      Login
    </button>
  </div>
  <h1>
    Logged Out
  </h1>
  <table
    class="debug-table"
  >
    <tbody>
      <tr>
        <td
          class="debug-itemName"
        >
          accessToken
        </td>
        <td
          class="debug-itemValue"
        >
          &lt;empty&gt;
        </td>
      </tr>
      <tr>
        <td
          class="debug-itemName"
        >
          refreshToken
        </td>
        <td
          class="debug-itemValue"
        >
          &lt;empty&gt;
        </td>
      </tr>
      <tr>
        <td
          class="debug-itemName"
        >
          state
        </td>
        <td
          class="debug-itemValue"
        >
          &lt;empty&gt;
        </td>
      </tr>
      <tr>
        <td
          class="debug-itemName"
        >
          code
        </td>
        <td
          class="debug-itemValue"
        >
          &lt;empty&gt;
        </td>
      </tr>
      <tr>
        <td
          class="debug-itemName"
        >
          codeVerifier
        </td>
        <td
          class="debug-itemValue"
        >
          &lt;empty&gt;
        </td>
      </tr>
      <tr>
        <td
          class="debug-itemName"
        >
          cookie
        </td>
        <td
          class="debug-itemValue"
        >
          &lt;empty&gt;
        </td>
      </tr>
      <tr>
        <td
          class="debug-itemName"
        >
          localStorage
        </td>
        <td
          class="debug-itemValue"
        >
          {}
        </td>
      </tr>
    </tbody>
  </table>
</main>
`;
